### ISSUE TITLE
Multi-channel convolution crashes network

### ISSUE THREAD

**Issue by swamiviv** [BVLC/caffe#4076]
As input to my deep network, I concatenate two data layers (each of them being image data in the form of RGB images) to get 6 channel data items. This is fed as input to a regular CNN architecture so essentially the first layer has conv. filters of depth 6.

1. When I start training in GPU mode, I find that the network blows up very soon and upon decreasing learning rate to very low values (1e-7) it still blows up but little later (after about 1000 iterations). But strangely, none of these happen in the CPU mode and the loss function promptly starts decreasing. But this is, of course, painfully slow.

2. When I put the data into hdf5 format and run it in GPU mode, I can run it successfully.

But ideally I would want to be able to do this using Image data layers since creating hdf5 data is time and memory consuming for abot 100K items.

So I am wondering if there is any issue with the way convolutions are performed on multiple channel data in GPU model. Am I missing something here ? Any help is appreciated.

**Comment by seanbell** [comment]
It sounds like you're describing regular "2D" convolution, i.e. the inputs and outputs are 4D with shape `(batch, channel, height, width)` -- what are the shapes of your blobs?

Are you using cuDNN?  If so, what version?

**Comment by swamiviv** [comment]
Yes, that was abuse of terminology on my part. For the particular experiment I did, the size was (64,6,100,100). I am using cuDNN-v3

**Comment by swamiviv** [comment]
My prototxt looks like this:

`
layer {
  name: "data1"
  type: "ImageData"
  top: "data1"
  top: "label"
  include {
    phase: TRAIN
  }
  image_data_param {
    source: <list1>
    batch_size: 64
  }
}
layer {
  name: "data2"
  type: "ImageData"
  top: "data2"
  top: "label2"
  include {
    phase: TRAIN
  }
  image_data_param {
    source: <list1>
    batch_size: 64
  }
 
layer{
name: "concat"
type: "Concat"
bottom: "data1"
bottom: "data2"
top: "data"
concat_param{
axis: 1
}
}
`

This data is then passed to the conv. layers. Each image in the list is of size 100x100x3.

**Comment by seanbell** [comment]
So, you're doing "2D" convolution (4D blobs), and not using the "ND" code path (where the blobs are 5D or more), right?

It's possible that the problem is with cuDNN.  Try running with the core caffe engine -- either recompile without cuDNN, or specify caffe as the engine inside each convolution layer (add `engine: CAFFE` inside `convolution_param`).  It could also be that your initial weights have a bad random distribution.

**Comment by swamiviv** [comment]
Yes "2D" convolution where the blobs are 4D but just more depth. Have edited the title.
Did a small check:

1. compiled with no cuDNN and no luck there.
2. tried xavier init. but no luck either. 

The strange thing is the same setting works for CPU mode or hdf5 inputs. So I am really puzzled.

**Comment by seanbell** [comment]
There are two possibilities: either (a) your learning problem is unstable or (b) the gradients are wrong.  Are there other custom layers somewhere in the network?  What's your loss function?

It looks like you're not subtracting any mean from your images.  That can make learning unstable.

It's hard to tell whether or not you have found a bug in caffe (vs a modeling problem, or some code you wrote).

**Comment by swamiviv** [comment]
Thanks for your inputs.

(a) your learning problem is unstable:
How does it explain the same network running in CPU mode or as hdf5 inputs with the same batch size ? 

(b) the gradients are wrong
Could be the case. Haven't verified it yet. But I am not using *any* custom code and yet I observe this behaviour.

(c) It looks like you're not subtracting any mean from your images
That was done before passing it to the network (since I used the same data as hdf5)

I am pretty sure its not a bug in caffe. But I am finding to hard to being to debug this issue. I will try writing the data as lmdbs and then do the same procedure.

**Comment by seanbell** [comment]
How did you subtract the mean if you're using an `ImageData` layer?  That layer reads normal images and outputs values in the range [0, 255].  You need to add a transform parameter to that layer, or use some other shift layer.

Good luck in debugging this issue.  Please follow up on the mailing list (debugging, usage, and modeling help have been moved over there).  Since there's a lot of noise on the Issues list, I'm going to close this issue; please re-open if you find a specific bug in caffe.

From https://github.com/BVLC/caffe/blob/master/CONTRIBUTING.md:
> _Please do not post usage, installation, or modeling questions, or other requests for help to Issues._
Use the [caffe-users list](https://groups.google.com/forum/#!forum/caffe-users) instead. This helps developers maintain a clear, uncluttered, and efficient view of the state of Caffe.

**Comment by cqwwww** [comment]
I would  like to know how you get your 6-channel images. I have met a similar problem of yours. And I have no idea how to put two 3-channel images into a 6-channel image.

**Comment by championway** [comment]
I met the same problem as you met, have you already solved this problem?
I would like to know the solution about multi-channel convolution, thanks.
